src/main/java/com/squareup/kotlinpoet/ParameterizedTypeName.kt

```diff
@@ -108,9 +112,43 @@ class ParameterizedTypeName internal constructor(
         get(ownerType, map = map).nestedClass(rawType.simpleName(), typeArguments) else
         ParameterizedTypeName(null, rawType, typeArguments)
     }
+
+    /** Returns a type name equivalent to type with given list of type arguments.  */
+    internal fun get(type: KClass<*>, nullable: Boolean, typeArguments: List<KTypeProjection>): TypeName {
+      if (typeArguments.isEmpty()) {
+        return type.asTypeName()
+      }
+
+      val enclosingClass = type.java.enclosingClass?.kotlin
+
+      return ParameterizedTypeName(
+          enclosingClass?.let {
+            get(it, false, typeArguments.drop(type.typeParameters.size))
+          },
+          type.asTypeName(),
+          typeArguments.take(type.typeParameters.size).map {
+            it.type?.asTypeName() ?: WildcardTypeName.subtypeOf(ANY)
+          },
+          nullable,
+          type.annotations.map { AnnotationSpec.get(it) })
+    }
   }
 }
 
 /** Returns a parameterized type equivalent to `type`.  */
 @JvmName("get")
 fun ParameterizedType.asParameterizedTypeName() = ParameterizedTypeName.get(this, mutableMapOf())
+
+/** Returns a class name equivalent to given Kotlin KType.  */
+fun KType.asTypeName(): TypeName {
+  val classifier = this.classifier
+  if (classifier is KTypeParameter) {
+    return classifier.asTypeVariableName()
+  }
+
+  if (classifier == null || classifier !is KClass<*>) {
```

Comment:
Can we test this code path?

Reply:
According to Kt documentation: "Returns `null` if this type is not denotable in Kotlin, for example if it is an intersection type.". So building the artificial example to test this path is not worth it: anyway, the result is not presentable in Kotlin :)